### PR TITLE
improve provider management

### PR DESCRIPTION
**Pull request by Gustry**
### What does it fix?
* Ticket: #4605 
* Funded by: DFAT
* Description: improve provider management

### Checklist:
- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Add to the changelog in metadata.txt if it's a new feature
- [ ] Unit test for new code added
- [ ] Request someone to review or test your PR

### CODE REVIEW

**Comment by Gustry** [comment]
Gentle ping from 2 days ago ;-)
Thanks, I need this commit for serialization

File: safe/gis/tools.py [hunk 1]
```diff
@@ -113,45 +129,194 @@ def reclassify_value(one_value, ranges):
...
+    :param layer_name: Layer name to use. Default to 'tmp'.
+    :type layer_name: basestring
+
+    :return: The layer or None if it's failed.
+    :rtype: QgsMapLayer
+    """
+    if provider in RASTER_DRIVERS:
+        return QgsRasterLayer(layer_uri, layer_name, provider)
+    elif provider in VECTOR_DRIVERS:
+        return QgsVectorLayer(layer_uri, layer_name, provider)
```
Comment (ismailsunni) [inline safe/gis/tools.py]: What about delimitedtext? I saw in the other part of code, we need to use it explicitly.
```
            # Explicitly use URI with delimiter or tests fail in Windows. TS.
            layer = QgsVectorLayer(
                'file:///%s?delimiter=,' % layer_uri, layer_name, driver)
```

File: safe/gis/tools.py [hunk 2]
```diff
@@ -113,45 +129,194 @@ def reclassify_value(one_value, ranges):
...
+            name = os.path.splitext(file_name)[0]
+        else:
+            # Might be a DB, take the DB name
+            source = QgsDataSourceURI(source)
+            name = source.table()
+
+    if not name:
+        name = 'default'
+
+    if qgis_version() >= 21800:
```
Comment (ismailsunni) [inline safe/gis/tools.py]: IIRC, you told me that we don't need to do this? Or is it the other API?

File: safe/gis/tools.py [hunk 3]
```diff
@@ -113,45 +129,194 @@ def reclassify_value(one_value, ranges):
...
+    if len(split) == 1:
+        return split[0], None
+    else:
+        return split
+
+
+def load_layer(full_layer_uri_string, name=None, provider=None):
+    """Helper to load and return a single QGIS layer based on our layer URI.
+
+    :param provider:
```
Comment (ismailsunni) [inline safe/gis/tools.py]: Docstring.

File: safe/gis/tools.py (hunk 2 above)
Comment (Gustry) [reply safe/gis/tools.py]: The getter is the same `name()` whatever the QGIS version, but the setter is different between QGIS 2.14 and 2.18. So it's fine.

File: safe/gis/tools.py (hunk 1 above)
Comment (Gustry) [reply safe/gis/tools.py]: Yes, I was thinking about this one. But this function is used when the caller of this function known the provider name. So if the user known the provider name, I expected the caller to have a well formatted  URI

File: safe/gui/tools/wizard/test/test_keyword_wizard.py [hunk 4]
```diff
@@ -2632,6 +2632,8 @@ def test_continuous_vector(self):
 
         self.assertDictEqual(real_keywords, expected_keyword)
 
+    @unittest.skip(
```
[the remark this answers is not in the document]
Comment (Gustry) [reply safe/gui/tools/wizard/test/test_keyword_wizard.py]: I forgot this one. I wanted to fix it.

File: safe/gis/tools.py (hunk 1 above)
Comment (ismailsunni) [reply safe/gis/tools.py]: ah I see. But perhaps put the expected URI for this special  case (delimited text) to make the future usage of this method easier.

**Pull request merged**